wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/nbt/NbtDelegates.kt

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qkl.wrapper.minecraft.nbt
+
+import net.minecraft.nbt.*
+import kotlin.properties.PropertyDelegateProvider
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KCallable
+import kotlin.reflect.KProperty
+
+/**
+ * A property delegate for NBT tags.
+ */
+public typealias NbtProperty<T> = ReadWriteProperty<Any?, T>
+
+/**
+ * A provider for NBT property delegates.
+ */
+public typealias NbtPropertyProvider<T> = PropertyDelegateProvider<Any?, NbtProperty<T>>
+
+/**
+ * A reference to a [Compound][NbtCompound]. To get one, create a variable and reference it with `::variableName`.
+ */
+public typealias CompoundProperty = KCallable<NbtCompound>
+
+internal class Delegate<T>(
+    val compound: CompoundProperty,
+    val name: String,
+    default: T? = null,
+    val toNbt: (T) -> NbtElement,
+    val fromNbt: (NbtElement) -> T,
+) : NbtProperty<T> {
+    init {
+        if (default != null) {
+            compound.call().put(name, toNbt(default))
+        }
+    }
+
+    override fun getValue(thisRef: Any?, property: KProperty<*>): T = fromNbt(compound.call().get(name)!!)
```

Comment:
Perhaps a syntax similar to that of map delegates (`val a: Int by mapOf('a' to 1)`) which obtains the correct key from property name would be preferable?

Reply:
I'm not sure what you mean. Values can be specified without explicit names when creating the delegates, and if a name isn't specified, the delegated variable's name will be used instead. The delegate used is meant to be internal.

Reply:
Ah, I see, you're doing the name choice in a delegate provider instead of in the delegate itself

That said, by providing overloads of `operator fun CompoundProperty.provideDelegate(thisRef: MyUI, prop: KProperty<X>)` for various X like Int, String, etc., it's possible to instead do `val someInt: Int by ::compound`